### PR TITLE
chore: upgrade to rust 1.66.0

### PR DESCRIPTION
**Pull request by lukesteensen**
I worked through a decent amount of the clippy changes, but ended up hitting an ICE. Pushing this now for visibility.

### CODE REVIEW

**Comment by fuchsnj** [comment]
I ran a full `cargo check` and `cargo build` on this without any ICE's. Did you try a `cargo clean` first?

**Comment by bruceg** [comment]
I updated this with current master and fixed a few more lints, but it still has a repeatable ICE when running `clippy`.

File: lib/value/src/kind/crud/remove.rs [hunk 1]
```diff
@@ -49,15 +49,13 @@ impl Kind {
...
-                                }
-                                Some(child) => child,
-                            }
-                            .remove_inner(&segments[1..], compact)
-                            .compact(object, field.clone(), compact)
+                            // The modified value is discarded here (It's not needed)
+                            object
+                                .known_mut()
+                                .get_mut(&Field::from(field.clone()))
+                                .map_or(&mut at_path_kind, |child| child)
```
Comment (bruceg) [inline lib/value/src/kind/crud/remove.rs:56]: Is this not `.unwrap_or(&mut at_path_kind)`?

File: lib/value/src/kind/crud/remove.rs [hunk 2]
```diff
@@ -104,15 +102,13 @@ impl Kind {
...
-                            }
-                            Some(child) => child,
-                        }
-                        .remove_inner(&segments[1..], compact)
-                        .compact(array, index as usize, compact)
+                        // The modified value is discarded here (It's not needed)
+                        array
+                            .known_mut()
+                            .get_mut(&(index as usize).into())
+                            .map_or(&mut at_path_kind, |child| child)
```
Comment (bruceg) [inline lib/value/src/kind/crud/remove.rs:109]: Ditto

File: lib/datadog/grok/src/grok.rs [hunk 3]
```diff
@@ -64,7 +64,7 @@ impl<'a> Iterator for MatchesIter<'a> {
         // that index, if anything.
         self.names.next().map(|(k, v)| {
             let key = k.as_str();
-            let value = self.captures.at(*v as usize).unwrap_or("");
+            let value = self.captures.at(*v).unwrap_or("");
```
Comment (bruceg) [inline lib/datadog/grok/src/grok.rs:67]: There are a lot of these conversions being dropped in this PR, but I don't see the variable types as having changed nor do I recall something in the Rust release notes that referenced it. What has enabled this change?

**Review by bruceg — APPROVED** [review]
LGTM aside from one possible pair of further simplifications and a question.

File: lib/datadog/grok/src/grok.rs (hunk 3 above)
Comment (lukesteensen) [reply lib/datadog/grok/src/grok.rs:67]: These were all unnecessary casts (i.e. `usize as usize`) that clippy now seems to detect.

File: lib/value/src/kind/crud/remove.rs (hunk 1 above)
Comment (lukesteensen) [reply lib/value/src/kind/crud/remove.rs:56]: Yep, good catch.

**Comment by jszwedko** [comment]
Seemingly blocked by https://github.com/dtolnay/inventory/issues/58

**Pull request merged**